ee/clickhouse/timer.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import uuid
+from collections import OrderedDict
+from functools import partial
+from threading import Condition, Thread
+from time import time
+from typing import Callable, Dict, Optional, Tuple
+
+from django.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+class TimerTask:
+    id: str
+    done: bool
+
+    def __init__(self, callback: Callable, *args, **kwargs):
+        self.callback = partial(callback, *args, **kwargs)
+        self.id = str(uuid.uuid4())
+        self.done = False
+
+    def run(self):
+        self.done = True
+        try:
+            self.callback()
+        except Exception as err:
+            logger.warn("TimerTask failed, ignoring error", err)
+
+
+class SingleThreadedTimer(Thread):
+    def __init__(self, timeout_ms: int, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.setDaemon(True)
+
+        self.timeout_ms = timeout_ms
+        self.started = False
+        self.lock = Condition()
+        self.tasks: OrderedDict = OrderedDict()
+
+    def schedule(self, callback: Callable, *args, **kwargs) -> TimerTask:
+        self.start()
+
+        with self.lock:
+            task = TimerTask(callback, *args, **kwargs)
+            self.tasks[task.id] = (task, time())
+            self.lock.notify()
+
+            return task
+
+    def cancel(self, task: TimerTask) -> None:
+        with self.lock:
+            del self.tasks[task.id]
+            self.lock.notify()
+
+    def start(self):
+        if not self.started:
+            self.started = True
+            super().start()
+
+    def run(self):
```

Comment:
and a comment here:

"""
Gets the scheduled tasks in FIFO order, and runs them after `timeout_ms` milliseconds.
"""

Reply:
Q: Why add an annotation here? Given this is threading-internal, user should never call this. Also the comment is oversimplifying things. I changed naming internally around a bit - hopefully the code is self-evident enough now.

Reply:
Makes more sense, yeah!